### ISSUE TITLE
Different TensorFlow Keras Code with same logic, tend to show very different behaviour.

### ISSUE THREAD

**Issue by abhaskumarsinha** [tensorflow/tensorflow#56742]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Source

binary

### Tensorflow Version

v2.9.0-18-gd8ce9f9c301 2.9.1

### Custom Code

No

### OS Platform and Distribution

Windows 11 64 Bit

### Python version

3.9.5

### CUDA/cuDNN version

11

### Current Behaviour?

I firstly apologize because despite days of pain, I'm still not very clear if its a bug or something which is keeping me from it. I'm not much experienced with me and really need someone experienced to guide me. Any clarification about status of bug or problem from my side is very much appreciated. 

So, I've two different codes for a DC GAN, which is basically a Generative Adverbial Neural Network using Cov Nets, despite similar model architecture and similar training steps, my code fails to converge in the way and speed of the second code. 

My Notebook (First version): https://colab.research.google.com/drive/12bMfjZgld6ETzPhPsz7HzoWHdL8056yG?usp=sharing
GFG Version (Second Version):  https://www.geeksforgeeks.org/deep-convolutional-gan-with-keras/

1. Results with my code vs the GFG Version of the Code.

GFG Version of Code:
  [1]: https://i.stack.imgur.com/k2UVN.png
*(GFG Version of the Code, after 1st Epoch and 128 batch size)*

My version of Code:
  [2]: https://i.stack.imgur.com/ciaAS.png

*(My Version of the Code, after training 1000 batch size and 60 epochs in total)*

As you can see, my model fails severely to converge to the desired result and for some reason, I notice mode collapse or something which doesn't lets my model converge to the desired results.

I'm posting the code of my training sequence as well as a bit of explaination of what it does in the following section:

```shell
def training_sequence(batch_size=1000, epochs=60):
    for epoch in range(epochs):
        print("Epoch: " + str(epoch + 1) + " of " + str(epochs))
        #Training Discriminator:

        GAN.layers[0].trainable = False
        GAN.layers[1].trainable = True

        #1 for real samples and 0 for fake samples
        fake_generated_examples = generator(generate_random_sequence(size = int(batch_size / 2)))
        zeros = tf.zeros((int(batch_size / 2), 1, 1, 1))

        #train discriminator
        discriminator.fit(fake_generated_examples, zeros)
        ones = tf.ones((int(batch_size / 2), 1, 1, 1))
        real_examples = x_train[int(batch_size / 2)*(epoch) : int(batch_size / 2)*(epoch + 1)].reshape(int(int(batch_size / 2)*(epoch + 1) - int(batch_size / 2)*(epoch)), 28, 28, 1)

        #train discriminator
        discriminator.fit(real_examples, ones)

        #Training Generator
        GAN.layers[0].trainable = True
        GAN.layers[1].trainable = False

        #train generator
        GAN.fit(generate_random_sequence(size = int(batch_size / 2)), ones)
```

There are two models (I'll post the summary just a bit after their introduction) - 
 1. Discriminator: Takes in the input of `(28, 28, 1)` and outputs: `[[[[1]]]]` for real samples (the ones from the dataset) and `[[[[0]]]]` for fake ones (which are generated from Generators.
 2. GAN: It is a Sequence of Generator and Discriminator models. It takes in a random value of `(1, 100)` and returns the output from the discriminator. 

`generate_random_sequences(size)` is just a function to generate a number of random 100 numbers of `size` rows and `(1,100)` columns which are fed into the generator.

`int(batch_size / 2)*(epoch)` used in the code is just used to select the first `batch_size` instances from the dataset in each epoch loop and keep selecting the next batch size and so on, so there is no repetition of selection of a data from dataset during each epoch and they remain unique.

Here's my model architecture:
  [3]: https://i.stack.imgur.com/2XeyJ.png

As you can see it's very similar to GFG Version of Models.

The training logic is similar in both cases, we train discriminator and then, GAN by locking discriminator by discriminator.trainable = False and training Generator this way. Despite training steps being similar for both models and same architecture, my model fails severely to converge as like the first one. I suppose its a problem with me, but even if its the case of bug, any guide or help would be very appreciated.

I'd refrain from using Gradient Tape from Tensorflow and stick to Keras for a while for simplicity.
```


### Standalone code to reproduce the issue

```shell
My Notebook: https://colab.research.google.com/drive/12bMfjZgld6ETzPhPsz7HzoWHdL8056yG?usp=sharing
GFG Version:  https://www.geeksforgeeks.org/deep-convolutional-gan-with-keras/
```

### Relevant log output

```shell
[1]: GFG Version
[2]: My version
  [1]: https://i.stack.imgur.com/k2UVN.png
  [2]: https://i.stack.imgur.com/ciaAS.png
```
</details>

**Comment by sushreebarsa** [comment]
@abhaskumarsinha 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by abhaskumarsinha** [comment]
The issue was spotted and fixed by the Keras-team. The issue was on my side, while training Models.
